source/dem/disable_particle_contact.cc

```diff
@@ -0,0 +1,335 @@
+  IndexSet       locally_relevant_dofs =
+    DoFTools::extract_locally_relevant_dofs(background_dh);
+
+  // Reinit all value of mobility at nodes
+  mobility_at_nodes.reinit(locally_owned_dofs,
+                           locally_relevant_dofs,
+                           mpi_communicator);
+  mobility_at_nodes = 0;
+
+  // Empty status (3) to nodes when no particle in cell
+  for (auto cell = active_ghost_cells.begin();
+       cell != active_ghost_cells.end();)
+    {
+      // Check if the cell has any particles
+      if (particle_handler.n_particles_in_cell(*cell) == 0)
+        {
+          std::vector<types::global_dof_index> local_dofs_indices(
+            dofs_per_cell);
+          (*cell)->get_dof_indices(local_dofs_indices);
+
+          // Remove empty cell from cells to reduce the number of check
+          cell = active_ghost_cells.erase(cell);
+
+          // Assign empty status to nodes
+          for (auto dof_index : local_dofs_indices)
+            {
+              mobility_at_nodes(dof_index) = mobility_status::empty;
+            }
+        }
+      else
+        {
+          ++cell;
+        }
+    }
+
+  mobility_at_nodes.update_ghost_values();
+
+  // Mobile status (2) to nodes (no overwrite of empty status) and to cell if
+  // the criteria is respected or cell has one or many empty nodes
+  // (empty neighbor)
+  for (auto cell = active_ghost_cells.begin();
+       cell != active_ghost_cells.end();)
+    {
+      // Assign mobility status to cell if has particles and
+      // - granular temperature > limit or
+      // - solid fraction of cell < limit or
+      // - is next to an empty cell
+      const unsigned int cell_id = (*cell)->active_cell_index();
+
+      std::vector<types::global_dof_index> local_dofs_indices(dofs_per_cell);
+      (*cell)->get_dof_indices(local_dofs_indices);
+
+      // Check if cell has an empty status node
+      bool has_empty_neighbor = false;
+      for (auto dof_index : local_dofs_indices)
+        {
+          if (mobility_at_nodes[dof_index] == mobility_status::empty)
+            {
+              has_empty_neighbor = true;
+              break;
+            }
+        }
+
+      if (granular_temperature_average[cell_id] >=
+            granular_temperature_threshold ||
+          solid_fractions[cell_id] <= solid_fraction_threshold ||
+          has_empty_neighbor)
+        {
+          // Insert cell in mobile status set
+          status_to_cell[mobility_status::mobile].insert(*cell);
+
+          // Remove cell from cell set
+          cell = active_ghost_cells.erase(cell);
+
+          // Assign mobile status to nodes
+          for (auto dof_index : local_dofs_indices)
+            {
+              // Don't overwrite empty nodes
+              mobility_at_nodes(dof_index) =
+                std::max((int)mobility_status::mobile,
+                         mobility_at_nodes[dof_index]);
+            }
+        }
+      else
+        {
+          ++cell;
+        }
+    }
+
```

Comment:
I feel like this member function is a bit too long. What would you think of breaking it down into 4 or even 5 separate members? It would be easier to maintain I think.

Reply:
Since I use a set of cells that I update during the loops, I can't really do that, otherwise I would have to return and use the updated set. It's a member function in the code write not, but in my next commit it would be a copy of the member (I would update this set only during load balancing)